usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_util.c

```diff
@@ -0,0 +1,947 @@
+	}
+}
+
+/*
+ * pqi_cmd_sm -- state machine for command
+ *
+ * NOTE: PQI_CMD_CMPLT and PQI_CMD_FATAL will drop the pd_mutex and regain
+ * it even if grab_lock==B_FALSE.
+ */
+void
+pqi_cmd_sm(pqi_cmd_t cmd, pqi_cmd_state_t new_state, boolean_t grab_lock)
+{
+	pqi_device_t	devp = cmd->pc_device;
+	pqi_state_t	s = cmd->pc_softc;
+
+	if (cmd->pc_softc->s_debug_level & DBG_LVL_STATE) {
+		cmn_err(CE_NOTE, "%s: cmd=%p (%s) -> (%s)\n", __func__,
+		    (void *)cmd, cmd_state_str(cmd->pc_cmd_state),
+		    cmd_state_str(new_state));
+	}
+	cmd->pc_last_state = cmd->pc_cmd_state;
+	cmd->pc_cmd_state = new_state;
+	switch (new_state) {
+	case PQI_CMD_UNINIT:
+		break;
+
+	case PQI_CMD_CONSTRUCT:
+		break;
+
+	case PQI_CMD_INIT:
+		break;
+
+	case PQI_CMD_QUEUED:
+		if (cmd->pc_last_state == PQI_CMD_STARTED)
+			break;
+		if (grab_lock == B_TRUE)
+			mutex_enter(&devp->pd_mutex);
+		cmd->pc_start_time = gethrtime();
+		cmd->pc_expiration = cmd->pc_start_time +
+		    ((hrtime_t)cmd->pc_pkt->pkt_time * NANOSEC);
+		devp->pd_active_cmds++;
+		atomic_inc_32(&s->s_cmd_queue_len);
+		list_insert_tail(&devp->pd_cmd_list, cmd);
+		if (grab_lock == B_TRUE)
+			mutex_exit(&devp->pd_mutex);
+		break;
+
+	case PQI_CMD_STARTED:
+		if (s->s_debug_level & (DBG_LVL_CDB | DBG_LVL_RQST))
+			pqi_dump_io(cmd->pc_io_rqst);
+		break;
+
+	case PQI_CMD_CMPLT:
+		if (grab_lock == B_TRUE)
+			mutex_enter(&devp->pd_mutex);
+
+		if ((cmd->pc_flags & PQI_FLAG_ABORTED) == 0) {
+			list_remove(&devp->pd_cmd_list, cmd);
+
+			devp->pd_active_cmds--;
+			atomic_dec_32(&s->s_cmd_queue_len);
+			pqi_free_io(cmd->pc_io_rqst);
+
+			cmd->pc_flags &= ~PQI_FLAG_FINISHING;
+			(void) ddi_taskq_dispatch(s->s_complete_taskq,
+			    cmd_finish_task, cmd, 0);
+		}
+
+		if (grab_lock == B_TRUE)
+			mutex_exit(&devp->pd_mutex);
+
+		break;
+
+	case PQI_CMD_FATAL:
+		if ((cmd->pc_last_state == PQI_CMD_QUEUED) ||
+		    (cmd->pc_last_state == PQI_CMD_STARTED)) {
+			if (grab_lock == B_TRUE)
+				mutex_enter(&devp->pd_mutex);
+
+			cmd->pc_flags |= PQI_FLAG_ABORTED;
+
+			/*
+			 * If this call came from aio_io_complete() when
+			 * dealing with a drive offline the flags will contain
+			 * PQI_FLAG_FINISHING so just clear it here to be
+			 * safe.
+			 */
+			cmd->pc_flags &= ~PQI_FLAG_FINISHING;
+
+			list_remove(&devp->pd_cmd_list, cmd);
+
+			devp->pd_active_cmds--;
+			atomic_dec_32(&s->s_cmd_queue_len);
+			if (cmd->pc_io_rqst)
+				pqi_free_io(cmd->pc_io_rqst);
+
+			(void) ddi_taskq_dispatch(s->s_complete_taskq,
+			    cmd_finish_task, cmd, 0);
+
+			if (grab_lock == B_TRUE)
+				mutex_exit(&devp->pd_mutex);
+		}
+		break;
+
+	case PQI_CMD_DESTRUCT:
+		if (grab_lock == B_TRUE)
+			mutex_enter(&devp->pd_mutex);
+
+		if (list_link_active(&cmd->pc_list)) {
+			list_remove(&devp->pd_cmd_list, cmd);
+			devp->pd_active_cmds--;
+			if (cmd->pc_io_rqst)
+				pqi_free_io(cmd->pc_io_rqst);
+		}
+
+		if (grab_lock == B_TRUE)
+			mutex_exit(&devp->pd_mutex);
+		break;
+
+	default:
+		/*
```

Comment:
This comment is describing a read of uninitialized memory poisoned by kmem.  Which is a real bug you should fix.

That said, I also don't see it being called with anything other than a constant value in this parameter, so perhaps the comment is just wrong.  Which you should also fix.